### PR TITLE
Show delegates names on transaction

### LINKED ISSUE #113: Show names of delegates / validators in transactions
In TransactionPanel the wallet should show the name of known delegates / validators on incoming / outgoing transactions so you do not have to look for them in delegates view

Maybe with current votes

### PR DESCRIPTION
**Pull request by phash**
Show the names of delegates on transactions (resolves #113)

HomePanel
TransactionPanel
TransactionDialog

### CODE REVIEW

File: src/main/java/org/semux/gui/dialog/TransactionDialog.java [hunk 1]
```diff
@@ -49,7 +53,13 @@ public TransactionDialog(JComponent parent, Transaction tx) {
         JLabel timestamp = new JLabel(SwingUtil.formatTimestamp(tx.getTimestamp()));
         JTextArea data = new JTextArea(Hex.PREF + Hex.encode(tx.getData()));
         data.setEditable(false);
-
+        JLabel delegateName = new JLabel();
+        delegateName.setVisible(false);
+        if (delegate.isPresent()) {
+            delegateName = new JLabel( delegate.get().getNameString());
+            delegateName.setVisible(true);
```
Comment (semuxproject) [inline src/main/java/org/semux/gui/dialog/TransactionDialog.java]:
```
delegateName.setText(...);
delegateName.setVisible(true);  // redundant
```

File: src/main/java/org/semux/gui/panel/HomePanel.java [hunk 2]
```diff
@@ -144,7 +147,7 @@ public HomePanel(Model model) {
     public static class TransactionPanel extends JPanel {
         private static final long serialVersionUID = 1L;
 
-        public TransactionPanel(Transaction tx, boolean inBound, boolean outBound) {
+        public TransactionPanel(Transaction tx, boolean inBound, boolean outBound, String delegatesName) {
```
Comment (semuxproject) [inline src/main/java/org/semux/gui/panel/HomePanel.java]: Why not using `Optinal<>` anymore?

File: src/main/java/org/semux/gui/panel/HomePanel.java [hunk 3]
```diff
@@ -209,7 +216,15 @@ public synchronized void actionPerformed(ActionEvent e) {
...
-
+    
+    private Optional<Delegate> getDelegateForTransaction(Transaction tx) {
+        Optional<Delegate> delegate= model.getDelegates().stream().filter(d -> Arrays.equals(d.getAddress(),tx.getFrom()) ).findFirst();
+        if (!delegate.isPresent()) {
+            delegate= model.getDelegates().stream().filter(d -> Arrays.equals( d.getAddress(),tx.getTo()) ).findFirst();
+        }
+        return delegate;
+    }
+
```
Comment (semuxproject) [inline src/main/java/org/semux/gui/panel/HomePanel.java]: Poorly written code. Consider merging the `OR` condition into the lambda expression, and adding documentation (rationality on both `from` and `to` are checked).  Also, your search is very slow if the number of delegates is large. Better use `DelegateState.getDelegate()`

File: src/main/java/org/semux/gui/panel/TransactionsPanel.java [hunk 4]
```diff
@@ -175,7 +180,15 @@ public synchronized void actionPerformed(ActionEvent e) {
...
-
+    
+    private Optional<Delegate> getDelegateForTransaction(Transaction tx) {
+        Optional<Delegate> delegate= model.getDelegates().stream().filter(d -> Arrays.equals(d.getAddress(),tx.getFrom()) ).findFirst();
+        if (!delegate.isPresent()) {
+            delegate= model.getDelegates().stream().filter(d -> Arrays.equals( d.getAddress(),tx.getTo()) ).findFirst();
+        }
+        return delegate;
+    }
+
```
Comment (semuxproject) [inline src/main/java/org/semux/gui/panel/TransactionsPanel.java]: Consider reusing previous defined method

File: src/main/java/org/semux/gui/panel/TransactionsPanel.java [hunk 5]
```diff
@@ -145,14 +147,17 @@ public Object getValueAt(int row, int column) {
...
                     from = Hex.encode(tx.getFrom());
                     from = accounts.containsKey(from) ? MessagesUtil.get("AccountNum") + accounts.get(from)
                             : Hex.PREF + from;
                 }
                 String to = Hex.encode(tx.getTo());
-                to = accounts.containsKey(to) ? MessagesUtil.get("AccountNum") + accounts.get(to) : Hex.PREF + to;
-                return from + " => " + to;
+                to = accounts.containsKey(to) ? MessagesUtil.get("AccountNum") + " #" + accounts.get(to)
+                        : Hex.PREF + to ;
+                return from + " => " + to + delegateName;
```
Comment (semuxproject) [inline src/main/java/org/semux/gui/panel/TransactionsPanel.java]: This column is `from => to`. Your addition is delegate name is very confusing!  Both from and to should be displayed with `delegate_name` or `account ##` or `address`.

**Comment by phash** [comment]
Better approach?

**Comment by testzcrypto** [comment]
Please use:
`JLabel(MessagesUtil.get("Delegate") + ":")`

Instead:
`JLabel("Delegate:")`

**Comment by phash** [comment]
done

**Pull request merged**